File: benchmarks/runtime/gate-b/julia/ekf_v1.jl
#!/usr/bin/env julia

# Persistent Julia control for the frozen Gate B EKF v1 workload.

using SHA
using Statistics

const ROOT = normpath(joinpath(@__DIR__, "..", "..", "..", ".."))
const TRACE_PATH = joinpath(ROOT, "benchmarks", "runtime", "gate-b", "ekf-input-v1.bin")
const EPISODE_LENGTH = 4096
const SCALED_INSTANCES = (1, 8, 64)
const DT = 0.05
const LANDMARK_X = 25.0
const LANDMARK_Y = -10.0
const QUANTIZATION = 1.0e-10
const REFERENCE_HASH = "ddca8ab17cb390839d4c77e7cecc5203122f249685f5a28c36fd342cf303a758"
const JULIA_HASH = "04c4f5b1b9b4af787525a4e45bf2270fa0653f43cfcb1d85792ed6977b6259ec"
const EXPECTED_STATE = [18.169827258925427, 4.339708695271022, 0.2557219366745068]
const EXPECTED_COVARIANCE = [
    0.3270953723043491 0.1509754472729972 -0.022618166436367253;
    0.1509754472729972 0.07105284175378412 -0.010486015657880304;
    -0.022618166436367253 -0.010486015657880304 0.0016395600302299483
]
const INITIAL_STATE = [2.0, 1.0, 0.15]
const INITIAL_COVARIANCE = [1.0 0.0 0.0; 0.0 1.0 0.0; 0.0 0.0 0.05]
const PROCESS_COVARIANCE = [0.04 0.0; 0.0 0.0025]
const MEASUREMENT_COVARIANCE = [0.25 0.0; 0.0 0.0009]

function load_trace()
    values = reinterpret(Float64, read(TRACE_PATH))
    length(values) == EPISODE_LENGTH * 4 || error("Gate B trace has the wrong length")
    reshape(copy(values), 4, EPISODE_LENGTH)
end

const TRACE = load_trace()

mutable struct Workspace
    instances::Int
    state::Matrix{Float64}
    covariance::Array{Float64,3}
    predicted_state::Vector{Float64}
    motion_jacobian::Matrix{Float64}
    control_jacobian::Matrix{Float64}
    gp::Matrix{Float64}
    predicted_covariance::Matrix{Float64}
    vq::Matrix{Float64}
    process_covariance::Matrix{Float64}
    measurement_jacobian::Matrix{Float64}
    hp::Matrix{Float64}
    innovation_covariance::Matrix{Float64}
    inverse_innovation::Matrix{Float64}
    pht::Matrix{Float64}
    gain::Matrix{Float64}
    correction::Vector{Float64}
    kh::Matrix{Float64}
    joseph_a::Matrix{Float64}
    ap::Matrix{Float64}
    corrected_covariance::Matrix{Float64}
    kr::Matrix{Float64}
    measurement_covariance::Matrix{Float64}
end

function Workspace(instances::Int)
    instances in SCALED_INSTANCES || error("instances must be 1, 8, or 64")
    ws = Workspace(
        instances, zeros(3, instances), zeros(3, 3, instances), zeros(3),
        zeros(3, 3), zeros(3, 2), zeros(3, 3), zeros(3, 3), zeros(3, 2),
        zeros(3, 3), zeros(2, 3), zeros(2, 3), zeros(2, 2), zeros(2, 2),
        zeros(3, 2), zeros(3, 2), zeros(3), zeros(3, 3), zeros(3, 3),
        zeros(3, 3), zeros(3, 3), zeros(3, 2), zeros(3, 3),
    )
    reset!(ws)
    ws
end

function reset!(ws::Workspace)
    @inbounds for instance in 1:ws.instances
        for row in 1:3
            ws.state[row, instance] = INITIAL_STATE[row]
            for column in 1:3
                ws.covariance[row, column, instance] = INITIAL_COVARIANCE[row, column]
            end
        end
    end
end

function matmul!(output, left, right)
    @inbounds for column in axes(right, 2), row in axes(left, 1)
        total = 0.0
        for inner in axes(left, 2)
            total += left[row, inner] * right[inner, column]
        end
        output[row, column] = total
    end
end

function matmul_right_transpose!(output, left, right)
    @inbounds for column in axes(right, 1), row in axes(left, 1)
        total = 0.0
        for inner in axes(left, 2)
            total += left[row, inner] * right[column, inner]
        end
        output[row, column] = total
    end
end

function step!(ws::Workspace, instance::Int, turn::Int)
    @inbounds begin
        velocity = TRACE[1, turn]
        angular_velocity = TRACE[2, turn]
        measured_range = TRACE[3, turn]
        measured_bearing = TRACE[4, turn]
        theta = ws.state[3, instance]
        cosine = cos(theta)
        sine = sin(theta)

        fill!(ws.motion_jacobian, 0.0)
        ws.motion_jacobian[1, 1] = 1.0
        ws.motion_jacobian[2, 2] = 1.0
        ws.motion_jacobian[3, 3] = 1.0
        ws.motion_jacobian[1, 3] = -velocity * sine * DT
        ws.motion_jacobian[2, 3] = velocity * cosine * DT
        fill!(ws.control_jacobian, 0.0)
        ws.control_jacobian[1, 1] = cosine * DT
        ws.control_jacobian[2, 1] = sine * DT
        ws.control_jacobian[3, 2] = DT
        ws.predicted_state[1] = ws.state[1, instance] + velocity * cosine * DT
        ws.predicted_state[2] = ws.state[2, instance] + velocity * sine * DT
        ws.predicted_state[3] = ws.state[3, instance] + angular_velocity * DT

        covariance = @view ws.covariance[:, :, instance]
        matmul!(ws.gp, ws.motion_jacobian, covariance)
        matmul_right_transpose!(ws.predicted_covariance, ws.gp, ws.motion_jacobian)
        matmul!(ws.vq, ws.control_jacobian, PROCESS_COVARIANCE)
        matmul_right_transpose!(ws.process_covariance, ws.vq, ws.control_jacobian)
        ws.predicted_covariance .+= ws.process_covariance

        delta_x = LANDMARK_X - ws.predicted_state[1]
        delta_y = LANDMARK_Y - ws.predicted_state[2]
        q = delta_x * delta_x + delta_y * delta_y
        q > 1.0e-12 || error("landmark distance")
        distance = sqrt(q)
        predicted_bearing = atan(delta_y, delta_x) - ws.predicted_state[3]
        ws.measurement_jacobian[1, 1] = -delta_x / distance
        ws.measurement_jacobian[2, 1] = delta_y / q
        ws.measurement_jacobian[1, 2] = -delta_y / distance
        ws.measurement_jacobian[2, 2] = -delta_x / q
        ws.measurement_jacobian[1, 3] = 0.0
        ws.measurement_jacobian[2, 3] = -1.0
        matmul!(ws.hp, ws.measurement_jacobian, ws.predicted_covariance)
        matmul_right_transpose!(ws.innovation_covariance, ws.hp, ws.measurement_jacobian)
        ws.innovation_covariance .+= MEASUREMENT_COVARIANCE
        determinant = ws.innovation_covariance[1, 1] * ws.innovation_covariance[2, 2] -
            ws.innovation_covariance[1, 2] * ws.innovation_covariance[2, 1]
        abs(determinant) > 1.0e-12 || error("innovation determinant")
        ws.inverse_innovation[1, 1] = ws.innovation_covariance[2, 2] / determinant
        ws.inverse_innovation[2, 1] = -ws.innovation_covariance[2, 1] / determinant
        ws.inverse_innovation[1, 2] = -ws.innovation_covariance[1, 2] / determinant
        ws.inverse_innovation[2, 2] = ws.innovation_covariance[1, 1] / determinant
        matmul_right_transpose!(ws.pht, ws.predicted_covariance, ws.measurement_jacobian)
        matmul!(ws.gain, ws.pht, ws.inverse_innovation)
        innovation_range = measured_range - distance
        innovation_bearing = measured_bearing - predicted_bearing
        for row in 1:3
            ws.correction[row] = ws.gain[row, 1] * innovation_range +
                ws.gain[row, 2] * innovation_bearing
            ws.state[row, instance] = ws.predicted_state[row] + ws.correction[row]
        end

        matmul!(ws.kh, ws.gain, ws.measurement_jacobian)
        for column in 1:3, row in 1:3
            ws.joseph_a[row, column] = (row == column ? 1.0 : 0.0) - ws.kh[row, column]
        end
        matmul!(ws.ap, ws.joseph_a, ws.predicted_covariance)
        matmul_right_transpose!(ws.corrected_covariance, ws.ap, ws.joseph_a)
        matmul!(ws.kr, ws.gain, MEASUREMENT_COVARIANCE)
        matmul_right_transpose!(ws.measurement_covariance, ws.kr, ws.gain)
        ws.corrected_covariance .+= ws.measurement_covariance
        for column in 1:3, row in 1:(column - 1)
            symmetric = 0.5 * (ws.corrected_covariance[row, column] +
                ws.corrected_covariance[column, row])
            ws.corrected_covariance[row, column] = symmetric
            ws.corrected_covariance[column, row] = symmetric
        end
        for column in 1:3, row in 1:3
            covariance[row, column] = ws.corrected_covariance[row, column]
        end
        for row in 1:3
            isfinite(ws.state[row, instance]) || error("non-finite state")
        end
        for column in 1:3, row in 1:3
            isfinite(covariance[row, column]) || error("non-finite covariance")
        end
        for diagonal in 1:3
            covariance[diagonal, diagonal] > 0.0 || error("covariance diagonal")
        end
        symmetry_error = 0.0
        for column in 1:3, row in 1:3
            symmetry_error = max(
                symmetry_error,
                abs(covariance[row, column] - covariance[column, row]),
            )
        end
        symmetry_error <= 1.0e-10 || error("covariance symmetry")
    end
end

function run_episode!(ws::Workspace)
    @inbounds for turn in 1:EPISODE_LENGTH, instance in 1:ws.instances
        step!(ws, instance, turn)
    end
end

function quantize(value::Float64)
    scaled = value / QUANTIZATION
    scaled >= 0.0 ? floor(Int64, scaled + 0.5) : ceil(Int64, scaled - 0.5)
end

function update_i64_le!(context, value::Int64, buffer::Vector{UInt8})
    bits = reinterpret(UInt64, value)
    @inbounds for index in 1:8
        buffer[index] = UInt8((bits >> (8 * (index - 1))) & 0xff)
    end
    SHA.update!(context, buffer)
end

function validate!()
    ws = Workspace(1)
    context = SHA.SHA256_CTX()
    buffer = zeros(UInt8, 8)
    for turn in 1:EPISODE_LENGTH
        step!(ws, 1, turn)
        for value in @view ws.state[:, 1]
            update_i64_le!(context, quantize(value), buffer)
        end
        for value in @view ws.covariance[:, :, 1]
            update_i64_le!(context, quantize(value), buffer)
        end
    end
    actual_hash = bytes2hex(SHA.digest!(context))
    actual_hash == JULIA_HASH || error("Julia trajectory hash mismatch: $actual_hash")
    maximum_error = max(
        maximum(abs.(ws.state[:, 1] .- EXPECTED_STATE)),
        maximum(abs.(ws.covariance[:, :, 1] .- EXPECTED_COVARIANCE)),
    )
    maximum_error <= 1.0e-10 || error("final-state error: $maximum_error")
    actual_hash
end

function benchmark(instances::Int, samples::Int; timeline::Bool=false)
    ws = Workspace(instances)
    reset!(ws)
    run_episode!(ws)
    durations = Float64[]
    allocated = Int[]
    gc_seconds = Float64[]
    for _ in 1:samples
        reset!(ws)
        result = @timed run_episode!(ws)
        push!(durations, result.time)
        push!(allocated, result.bytes)
        push!(gc_seconds, result.gctime)
    end
    if timeline
        for sample in eachindex(durations)
            println(
                "{\"lane\":\"julia-persistent\",\"sample\":$(sample - 1)," *
                "\"turns\":$EPISODE_LENGTH,\"elapsed_ns\":$(round(Int, durations[sample] * 1.0e9))," *
                "\"gc_ns\":$(round(Int, gc_seconds[sample] * 1.0e9))," *
                "\"allocated_bytes\":$(allocated[sample])}"
            )
        end
        return
    end
    order = sortperm(durations)
    middle = order[(samples + 1) ÷ 2]
    println(join((
        "julia-persistent", instances, samples,
        durations[middle] * 1000.0,
        minimum(durations) * 1000.0,
        maximum(durations) * 1000.0,
        allocated[middle], gc_seconds[middle] * 1000.0,
    ), ','))
end

function main(arguments)
    samples = 9
    instances = collect(SCALED_INSTANCES)
    timeline = false
    index = 1
    while index <= length(arguments)
        if arguments[index] == "--samples"
            index += 1
            samples = parse(Int, arguments[index])
        elseif arguments[index] == "--instances"
            index += 1
            instances = [parse(Int, arguments[index])]
        elseif arguments[index] == "--timeline"
            timeline = true
        elseif arguments[index] == "--self-test"
            println("julia_hash=$(validate!()) reference_hash=$REFERENCE_HASH")
            return
        else
            error("unknown argument: $(arguments[index])")
        end
        index += 1
    end
    samples > 0 || error("samples must be positive")
    timeline || println("runtime,instances,samples,median_ms,min_ms,max_ms,allocated_bytes,gc_ms")
    diagnostic_hash = validate!()
    for count in instances
        benchmark(count, samples; timeline=timeline)
    end
    println(stderr, "validated trajectory $diagnostic_hash (reference $REFERENCE_HASH) on Julia $(VERSION), threads=$(Threads.nthreads())")
end

main(ARGS)

File: benchmarks/runtime/gate-b/numpy/ekf_v1.py
#!/usr/bin/env python3
"""Persistent NumPy control for the frozen Gate B EKF v1 workload."""

from __future__ import annotations

import argparse
import contextlib
import gc
import hashlib
import importlib.util
import io
import json
import math
import os
import platform
import struct
import sys
import time
import warnings
from pathlib import Path
from typing import Any

import numpy as np


ROOT = Path(__file__).resolve().parents[4]
TRACE_PATH = ROOT / "benchmarks/runtime/gate-b/ekf-input-v1.bin"
MANIFEST_PATH = ROOT / "benchmarks/runtime/gate-b/ekf-v1.json"
GENERATOR_PATH = ROOT / "scripts/generate-gate-b-ekf-trace.py"

EPISODE_LENGTH = 4_096
SCALED_INSTANCES = (1, 8, 64)
DT = 0.05
LANDMARK_X = 25.0
LANDMARK_Y = -10.0
ABSOLUTE_TOLERANCE = 1.0e-10
RELATIVE_TOLERANCE = 1.0e-10
QUANTIZATION = 1.0e-10
INITIAL_STATE = np.array([2.0, 1.0, 0.15], dtype=np.float64, order="F")
INITIAL_COVARIANCE = np.array(
    [[1.0, 0.0, 0.0], [0.0, 1.0, 0.0], [0.0, 0.0, 0.05]],
    dtype=np.float64,
    order="F",
)
PROCESS_COVARIANCE = np.array(
    [[0.04, 0.0], [0.0, 0.0025]], dtype=np.float64, order="F"
)
MEASUREMENT_COVARIANCE = np.array(
    [[0.25, 0.0], [0.0, 0.0009]], dtype=np.float64, order="F"
)
IDENTITY = np.eye(3, dtype=np.float64, order="F")


def _load_trace() -> np.ndarray:
    values = np.fromfile(TRACE_PATH, dtype="<f8")
    if values.size != EPISODE_LENGTH * 4:
        raise RuntimeError("Gate B trace has the wrong number of f64 values")
    trace = np.array(values.reshape((EPISODE_LENGTH, 4)), dtype=np.float64, order="F")
    if not trace.flags.f_contiguous:
        raise RuntimeError("Gate B NumPy trace must be Fortran contiguous")
    return trace


def _load_manifest() -> dict[str, Any]:
    with MANIFEST_PATH.open("r", encoding="utf-8") as handle:
        return json.load(handle)


TRACE = _load_trace()
MANIFEST = _load_manifest()


class Workspace:
    """All mutable state and matrix scratch for one instance scale."""

    def __init__(self, instances: int) -> None:
        if instances not in SCALED_INSTANCES:
            raise ValueError("instances must be one of 1, 8, or 64")
        self.instances = instances
        self.state = np.empty((3, instances), dtype=np.float64, order="F")
        self.covariance = np.empty(
            (3, 3, instances), dtype=np.float64, order="F"
        )
        self.predicted_state = np.empty(3, dtype=np.float64, order="F")
        self.corrected_state = np.empty(3, dtype=np.float64, order="F")
        self.motion_jacobian = np.empty((3, 3), dtype=np.float64, order="F")
        self.control_jacobian = np.empty((3, 2), dtype=np.float64, order="F")
        self.gp = np.empty((3, 3), dtype=np.float64, order="F")
        self.predicted_covariance = np.empty(
            (3, 3), dtype=np.float64, order="F"
        )
        self.vq = np.empty((3, 2), dtype=np.float64, order="F")
        self.process_covariance = np.empty((3, 3), dtype=np.float64, order="F")
        self.measurement_jacobian = np.empty(
            (2, 3), dtype=np.float64, order="F"
        )
        self.hp = np.empty((2, 3), dtype=np.float64, order="F")
        self.innovation_covariance = np.empty(
            (2, 2), dtype=np.float64, order="F"
        )
        self.inverse_innovation = np.empty((2, 2), dtype=np.float64, order="F")
        self.pht = np.empty((3, 2), dtype=np.float64, order="F")
        self.gain = np.empty((3, 2), dtype=np.float64, order="F")
        self.innovation = np.empty(2, dtype=np.float64, order="F")
        self.correction = np.empty(3, dtype=np.float64, order="F")
        self.kh = np.empty((3, 3), dtype=np.float64, order="F")
        self.joseph_a = np.empty((3, 3), dtype=np.float64, order="F")
        self.ap = np.empty((3, 3), dtype=np.float64, order="F")
        self.corrected_covariance = np.empty(
            (3, 3), dtype=np.float64, order="F"
        )
        self.kr = np.empty((3, 2), dtype=np.float64, order="F")
        self.measurement_covariance = np.empty(
            (3, 3), dtype=np.float64, order="F"
        )
        self.reset()

    def reset(self) -> None:
        self.state[:] = INITIAL_STATE[:, None]
        for instance in range(self.instances):
            self.covariance[:, :, instance] = INITIAL_COVARIANCE

    def step(self, instance: int, row: np.ndarray) -> None:
        state = self.state[:, instance]
        covariance = self.covariance[:, :, instance]
        velocity = float(row[0])
        angular_velocity = float(row[1])
        measured_range = float(row[2])
        measured_bearing = float(row[3])
        theta = float(state[2])
        cosine = math.cos(theta)
        sine = math.sin(theta)

        self.motion_jacobian[:] = IDENTITY
        self.motion_jacobian[0, 2] = -velocity * sine * DT
        self.motion_jacobian[1, 2] = velocity * cosine * DT
        self.control_jacobian.fill(0.0)
        self.control_jacobian[0, 0] = cosine * DT
        self.control_jacobian[1, 0] = sine * DT
        self.control_jacobian[2, 1] = DT
        self.predicted_state[0] = state[0] + velocity * cosine * DT
        self.predicted_state[1] = state[1] + velocity * sine * DT
        self.predicted_state[2] = state[2] + angular_velocity * DT

        np.matmul(self.motion_jacobian, covariance, out=self.gp)
        np.matmul(
            self.gp, self.motion_jacobian.T, out=self.predicted_covariance
        )
        np.matmul(self.control_jacobian, PROCESS_COVARIANCE, out=self.vq)
        np.matmul(
            self.vq, self.control_jacobian.T, out=self.process_covariance
        )
        np.add(
            self.predicted_covariance,
            self.process_covariance,
            out=self.predicted_covariance,
        )

        delta_x = LANDMARK_X - self.predicted_state[0]
        delta_y = LANDMARK_Y - self.predicted_state[1]
        q = delta_x * delta_x + delta_y * delta_y
        if q <= 1.0e-12:
            raise ArithmeticError("landmark distance")
        distance = math.sqrt(q)
        predicted_bearing = (
            math.atan2(delta_y, delta_x) - self.predicted_state[2]
        )
        self.measurement_jacobian[0, 0] = -delta_x / distance
        self.measurement_jacobian[1, 0] = delta_y / q
        self.measurement_jacobian[0, 1] = -delta_y / distance
        self.measurement_jacobian[1, 1] = -delta_x / q
        self.measurement_jacobian[0, 2] = 0.0
        self.measurement_jacobian[1, 2] = -1.0
        np.matmul(
            self.measurement_jacobian,
            self.predicted_covariance,
            out=self.hp,
        )
        np.matmul(
            self.hp,
            self.measurement_jacobian.T,
            out=self.innovation_covariance,
        )
        np.add(
            self.innovation_covariance,
            MEASUREMENT_COVARIANCE,
            out=self.innovation_covariance,
        )
        determinant = (
            self.innovation_covariance[0, 0]
            * self.innovation_covariance[1, 1]
            - self.innovation_covariance[0, 1]
            * self.innovation_covariance[1, 0]
        )
        if abs(determinant) <= 1.0e-12:
            raise ArithmeticError("innovation determinant")
        self.inverse_innovation[0, 0] = (
            self.innovation_covariance[1, 1] / determinant
        )
        self.inverse_innovation[1, 0] = (
            -self.innovation_covariance[1, 0] / determinant
        )
        self.inverse_innovation[0, 1] = (
            -self.innovation_covariance[0, 1] / determinant
        )
        self.inverse_innovation[1, 1] = (
            self.innovation_covariance[0, 0] / determinant
        )
        np.matmul(
            self.predicted_covariance,
            self.measurement_jacobian.T,
            out=self.pht,
        )
        np.matmul(self.pht, self.inverse_innovation, out=self.gain)
        self.innovation[0] = measured_range - distance
        self.innovation[1] = measured_bearing - predicted_bearing
        np.matmul(self.gain, self.innovation, out=self.correction)
        np.add(
            self.predicted_state, self.correction, out=self.corrected_state
        )

        np.matmul(self.gain, self.measurement_jacobian, out=self.kh)
        np.subtract(IDENTITY, self.kh, out=self.joseph_a)
        np.matmul(
            self.joseph_a, self.predicted_covariance, out=self.ap
        )
        np.matmul(
            self.ap, self.joseph_a.T, out=self.corrected_covariance
        )
        np.matmul(self.gain, MEASUREMENT_COVARIANCE, out=self.kr)
        np.matmul(
            self.kr, self.gain.T, out=self.measurement_covariance
        )
        np.add(
            self.corrected_covariance,
            self.measurement_covariance,
            out=self.corrected_covariance,
        )
        for column in range(3):
            for matrix_row in range(column):
                symmetric = 0.5 * (
                    self.corrected_covariance[matrix_row, column]
                    + self.corrected_covariance[column, matrix_row]
                )
                self.corrected_covariance[matrix_row, column] = symmetric
                self.corrected_covariance[column, matrix_row] = symmetric

        if not np.isfinite(self.corrected_state).all():
            raise ArithmeticError("non-finite state")
        if not np.isfinite(self.corrected_covariance).all():
            raise ArithmeticError("non-finite covariance")
        if not all(
            self.corrected_covariance[index, index] > 0.0
            for index in range(3)
        ):
            raise ArithmeticError("covariance diagonal")
        symmetry_error = 0.0
        for column in range(3):
            for matrix_row in range(3):
                symmetry_error = max(
                    symmetry_error,
                    abs(
                        self.corrected_covariance[matrix_row, column]
                        - self.corrected_covariance[column, matrix_row]
                    ),
                )
        if symmetry_error > 1.0e-10:
            raise ArithmeticError("covariance symmetry")

        state[:] = self.corrected_state
        covariance[:] = self.corrected_covariance

    def run_episode(self) -> None:
        for turn in range(EPISODE_LENGTH):
            row = TRACE[turn]
            for instance in range(self.instances):
                self.step(instance, row)


WORKSPACES = {instances: Workspace(instances) for instances in SCALED_INSTANCES}


def _reference_module() -> Any:
    spec = importlib.util.spec_from_file_location(
        "gate_b_trace_generator", GENERATOR_PATH
    )
    if spec is None or spec.loader is None:
        raise RuntimeError("cannot load Gate B scalar oracle")
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)
    return module


def _close(actual: float, expected: float) -> bool:
    tolerance = ABSOLUTE_TOLERANCE + RELATIVE_TOLERANCE * abs(expected)
    return abs(actual - expected) <= tolerance


def _quantize(value: float) -> int:
    scaled = value / QUANTIZATION
    if scaled >= 0.0:
        return math.floor(scaled + 0.5)
    return math.ceil(scaled - 0.5)


def validate_per_turn(workspace: Workspace) -> str:
    oracle = _reference_module()
    workspace.reset()
    expected_state = oracle.INITIAL_STATE
    expected_covariance = oracle.INITIAL_COVARIANCE
    digest = hashlib.sha256()
    for turn in range(EPISODE_LENGTH):
        expected_state, expected_covariance, _, _, _ = oracle.ekf_step(
            expected_state,
            expected_covariance,
            tuple(float(value) for value in TRACE[turn]),
        )
        expected_values = expected_state + expected_covariance
        for instance in range(workspace.instances):
            workspace.step(instance, TRACE[turn])
            actual_values = tuple(
                float(value) for value in workspace.state[:, instance]
            )
            actual_values += tuple(
                float(value)
                for value in workspace.covariance[:, :, instance].reshape(
                    -1, order="F"
                )
            )
            for index, (actual, expected) in enumerate(
                zip(actual_values, expected_values)
            ):
                if not _close(actual, expected):
                    raise AssertionError(
                        f"Gate B NumPy mismatch at turn {turn + 1}, "
                        f"instance {instance}, value {index}: "
                        f"{actual} != {expected}"
                    )
                if instance == 0:
                    digest.update(struct.pack("<q", _quantize(actual)))
    result = digest.hexdigest()
    return result


def benchmark(instances: int, samples: int) -> dict[str, Any]:
    if samples < 1:
        raise ValueError("samples must be positive")
    workspace = WORKSPACES[instances]
    diagnostic_hash = validate_per_turn(workspace)
    workspace.reset()
    workspace.run_episode()
    durations: list[int] = []
    gc_durations: list[int] = []
    gc_started_ns = 0
    gc_total_ns = 0

    def gc_event(phase: str, _info: dict[str, int]) -> None:
        nonlocal gc_started_ns, gc_total_ns
        if phase == "start":
            gc_started_ns = time.perf_counter_ns()
        elif gc_started_ns:
            gc_total_ns += time.perf_counter_ns() - gc_started_ns
            gc_started_ns = 0

    gc.callbacks.append(gc_event)
    try:
        for _ in range(samples):
            workspace.reset()
            gc_before = gc_total_ns
            started = time.perf_counter_ns()
            workspace.run_episode()
            durations.append(time.perf_counter_ns() - started)
            gc_durations.append(gc_total_ns - gc_before)
    finally:
        gc.callbacks.remove(gc_event)
    return {
        "type": "benchmark-result",
        "lane": "numpy-persistent",
        "instances": instances,
        "turns": EPISODE_LENGTH,
        "samples_ns": durations,
        "gc_samples_ns": gc_durations,
        "allocation_count": None,
        "allocated_bytes": None,
        "correctness": True,
        "quantized_state_hash": diagnostic_hash,
        "reference_quantized_state_hash": MANIFEST["reference"][
            "quantized_trajectory_sha256"
        ],
    }


def describe() -> dict[str, Any]:
    config_output = io.StringIO()
    with warnings.catch_warnings(), contextlib.redirect_stdout(config_output):
        warnings.simplefilter("ignore", UserWarning)
        np.show_config()
    numpy_config = config_output.getvalue().strip()
    config = getattr(np.__config__, "CONFIG", {})
    dependencies = config.get("Build Dependencies", {})
    provider_names = {
        dependency.get("name", "unknown")
        for kind, dependency in dependencies.items()
        if kind.lower() in {"blas", "lapack"}
    }
    if provider_names:
        blas_lapack_provider = ", ".join(sorted(provider_names))
    else:
        lowered_config = numpy_config.lower()
        blas_lapack_provider = next(
            (
                provider
                for provider in ("Accelerate", "OpenBLAS", "MKL", "BLIS")
                if provider.lower() in lowered_config
            ),
            "unknown",
        )
    return {
        "type": "ready",
        "protocol": "gate-b-numpy-v1",
        "pid": os.getpid(),
        "python": platform.python_version(),
        "numpy": np.__version__,
        "numpy_config": numpy_config,
        "blas_lapack_provider": blas_lapack_provider,
        "trace_sha256": MANIFEST["trace"]["sha256"],
        "workload": MANIFEST["workload"],
        "episode_length": EPISODE_LENGTH,
        "instances": list(SCALED_INSTANCES),
        "fortran_contiguous": {
            "trace": bool(TRACE.flags.f_contiguous),
            "state": all(
                workspace.state.flags.f_contiguous
                for workspace in WORKSPACES.values()
            ),
            "covariance": all(
                workspace.covariance.flags.f_contiguous
                for workspace in WORKSPACES.values()
            ),
        },
    }


def emit(payload: dict[str, Any]) -> None:
    print(json.dumps(payload, sort_keys=True), flush=True)


def serve() -> int:
    emit(describe())
    for line in sys.stdin:
        try:
            request = json.loads(line)
            command = request.get("command")
            if command == "describe":
                emit(describe())
            elif command == "benchmark":
                emit(
                    benchmark(
                        int(request["instances"]), int(request["samples"])
                    )
                )
            elif command == "quit":
                emit({"type": "bye"})
                return 0
            else:
                raise ValueError(f"unknown command: {command!r}")
        except Exception as error:  # Worker errors must remain structured.
            emit(
                {
                    "type": "error",
                    "error": type(error).__name__,
                    "message": str(error),
                }
            )
    return 0


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--self-test",
        action="store_true",
        help="validate one EKF outside the benchmark protocol",
    )
    args = parser.parse_args()
    if args.self_test:
        emit(
            {
                "type": "self-test",
                "correctness": True,
                "quantized_state_hash": validate_per_turn(WORKSPACES[1]),
                "reference_quantized_state_hash": MANIFEST["reference"][
                    "quantized_trajectory_sha256"
                ],
            }
        )
        return 0
    return serve()


if __name__ == "__main__":
    raise SystemExit(main())
